includes/wikia/services/RevisionUpvotesService.class.php

```diff
@@ -0,0 +1,290 @@
+			$upvoteId = $this->addUpvoteRevision( $wikiId, $pageId, $revisionId, $userId );
+		}
+
+		if ( !empty( $upvoteId ) ) {
+			$this->addUserUpvote( $upvoteId, $userId, $fromUser );
+		}
+	}
+
+	/**
+	 * Remove upvote for given revision made by user
+	 *
+	 * @param int $id
+	 * @param int $fromUser
+	 */
+	public function removeUpvote( $id, $fromUser ) {
+		$db = $this->getDatabaseForWrite();
+
+		$db->begin();
+
+		( new \WikiaSQL() )
+			->DELETE( self::UPVOTE_TABLE )
+			->WHERE( 'id' )->EQUAL_TO( $id )
+			->AND_( 'from_user' )->EQUAL_TO( $fromUser )
+			->run( $db );
+
+		( new \WikiaSQL() )
+			->UPDATE( self::UPVOTE_USERS_TABLE )
+			->SET_RAW( 'total', 'IF(`total` > 0, `total` - 1, 0)' )
+			->SET_RAW( 'new', 'IF(`new` > 0, `new` - 1, 0)' )
+			->WHERE( 'id' )->EQUAL_TO( $id )
+			->AND_( 'from_user' )->EQUAL_TO( $fromUser )
+			->run( $db );
+
+		$db->commit();
+	}
+
+	/**
+	 * Get all upvotes for given revision
+	 *
+	 * @param int $wikiId
+	 * @param int $revisionId
+	 * @return bool|array
+	 */
+	public function getRevisionUpvotes( $wikiId, $revisionId ) {
+		$db = $this->getDatabaseForRead();
+
+		$upvote = ( new \WikiaSQL() )
+			->SELECT_ALL()
+			->FROM( self::UPVOTE_REVISIONS_TABLE )->AS_( 'revs' )
+			->LEFT_JOIN( self::UPVOTE_TABLE )->AS_( 'uv' )
+			->ON( 'revs.upvote_id', 'uv.upvote_id' )
+			->WHERE( 'wiki_id' )->EQUAL_TO( $wikiId )
+			->AND_( 'revision_id' )->EQUAL_TO( $revisionId )
+			->runLoop( $db, function( &$upvote, $row ) {
+				if ( empty( $upvote['revision'] ) ) {
+					$upvote['revision'] = [
+						'wikiId' => $row->wiki_id,
+						'pageId' => $row->page_id,
+						'revisionId' => $row->revision_id,
+						'userId' => $row->user_id,
+						'upvoteId' => $row->upvote_id
+					];
+				}
+
+				$upvote['upvotes'][] = [
+					'from_user' => $row->from_user
+				];
+			} );
+
+		$upvote['count'] = count( $upvote['upvotes'] );
+
+		return $upvote;
+	}
+
+	/**
+	 * Get upvotes data for many revisions
+	 *
+	 * @param int $wikiId
+	 * @param array $revisionsIds
+	 * @return bool|array
+	 */
+	public function getRevisionsUpvotes( $wikiId, array $revisionsIds ) {
+		$db = $this->getDatabaseForRead();
+
+		$upvotes = ( new \WikiaSQL() )
+			->SELECT_ALL()
+			->FROM( self::UPVOTE_REVISIONS_TABLE )->AS_( 'revs' )
+			->LEFT_JOIN( self::UPVOTE_TABLE )->AS_( 'uv' )
+			->ON( 'revs.upvote_id', 'uv.upvote_id' )
+			->WHERE( 'wiki_id' )->EQUAL_TO( $wikiId )
+			->AND_( 'revision_id' )->IN( $revisionsIds )
+			->runLoop( $db, function( &$upvotes, $row ) {
+				if ( empty( $upvotes[$row->revision_id]['revision'] ) ) {
+					$upvotes[$row->revision_id]['revision'] = [
+						'wikiId' => $row->wiki_id,
+						'pageId' => $row->page_id,
+						'revisionId' => $row->revision_id,
+						'userId' => $row->user_id,
+						'upvoteId' => $row->upvote_id
+					];
+				}
+
+				$upvotes[$row->revision_id]['upvotes'][] = [
+					'from_user' => $row->from_user
+				];
+			} );
+
+		foreach( $upvotes as $revisionId => $upvote ) {
```

Comment:
Minor: space needed between the `foreach` and the opening parenthesis per coding conventions.